src/consensus/role_leader.rs

```diff
@@ -74,74 +74,81 @@ impl LeaderRole for Consensus {
-        });
+        // If we already have a consensusproposal for this slot, then we voted on it,
+        // and so we must repropose it (in case a commit was reached somewhere)
+        if self.bft_round_state.current_proposal.slot == self.bft_round_state.slot {
+            debug!("♻️ Starting new view with the same ConsensusProposal as previous views")
+        } else {
+            // TODO: keep candidates around?
+            let mut new_validators_to_bond = std::mem::take(&mut self.validator_candidates);
+            new_validators_to_bond.retain(|v| {
+                self.bft_round_state
+                    .staking
+                    .get_stake(&v.pubkey)
+                    .unwrap_or(0)
+                    > MIN_STAKE
+                    && !self.bft_round_state.staking.is_bonded(&v.pubkey)
+            });
 
-        debug!(
-            "🚀 Starting new slot {} (view {}) with {} existing validators and {} candidates",
-            self.bft_round_state.slot,
-            self.bft_round_state.view,
-            self.bft_round_state.staking.bonded().len(),
-            new_validators_to_bond.len()
-        );
+            debug!(
+                "🚀 Starting new slot {} (view {}) with {} existing validators and {} candidates",
+                self.bft_round_state.slot,
+                self.bft_round_state.view,
+                self.bft_round_state.staking.bonded().len(),
+                new_validators_to_bond.len()
+            );
 
-        // Creates ConsensusProposal
-        // Query new cut to Mempool
-        trace!(
-            "Querying Mempool for a new cut with Staking: {:#?}",
-            self.bft_round_state.staking
-        );
+            // Creates ConsensusProposal
+            // Query new cut to Mempool
+            trace!(
+                "Querying Mempool for a new cut with Staking: {:#?}",
+                self.bft_round_state.staking
+            );
 
-        let cut = match tokio::time::timeout(
-            std::time::Duration::from_millis(self.config.consensus.slot_duration),
-            self.bus
-                .request(QueryNewCut(self.bft_round_state.staking.clone())),
-        )
-        .await
-        .context("Timeout while querying Mempool")
-        {
-            Ok(Ok(cut)) => cut,
-            Ok(Err(err)) | Err(err) => {
-                // In case of an error, we reuse the last cut to avoid being considered byzantine
-                error!(
-                    "Could not get a new cut from Mempool {:?}. Reusing previous one...",
-                    err
-                );
-                self.bft_round_state.last_cut_seen.clone()
+            let cut = match tokio::time::timeout(
+                std::time::Duration::from_millis(self.config.consensus.slot_duration),
+                self.bus
+                    .request(QueryNewCut(self.bft_round_state.staking.clone())),
+            )
+            .await
+            .context("Timeout while querying Mempool")
```

Comment:
The QueryNewCut handler in mempool is a sync receiver, I don't think it is usefull to have this timeout around QueryNewCut. We are testing the bus there, and if the bus is slow to answer the new cut, it will be slow to send the messages to the network layer... I think we are adding complexity here where it's not necessary

A better solution could be to add "priority" over bus messages to handle QueryNewCut before the others if we think it can be an issue

Did you encounter a case where the QueryNewCut was slow ?